### PR TITLE
New Job latency

### PR DESCRIPTION
**Pull request by Shourya742**
This PR adds:
* SV2 New Job Latency (Pool): Prevhash
* SV2 New Job Latency (JDC): Prevhash
* SV1 New Job Latency
* SV2 New Job Latency (JDC)
* SV2 New Job Latency (Pool)

### CODE REVIEW

File: docker-compose-config-a.yaml [hunk 1]
```diff
@@ -211,7 +253,46 @@ services:
...
+      - PROM_ADDRESS=10.5.0.22:5677
+      - PROXY_TYPE=tp-pool
+    container_name: sv2-tp-pool-proxy
+    depends_on:
+      - sv2-custom-proxy-builder
+      - template-provider-pool-side
+    restart: unless-stopped
+    networks:
+      sv2-net:
+        ipv4_address: 10.5.0.22
```
Comment (GitGab19) [inline docker-compose-config-a.yaml]: Why is needed a tp-pool proxy in this config?

File: docker-compose-config-a.yaml [hunk 2]
```diff
@@ -211,7 +253,46 @@ services:
...
+
+  sv2-tranlator-node-proxy:
+    image: sv1-custom-proxy-builder-image
+    command: ["./target/release/sv1-custom-proxy"]
+    ports:
+      - "34255:34255"
+      - "5676:5676"
+    environment:
+      - SERVER=0.0.0.0:34255
+      - CLIENT=10.5.0.7:34256
```
Comment (GitGab19) [inline docker-compose-config-a.yaml:266]: Is this proxy located between translator and miners, right?
If so, shouldn't be `CLIENT=10.5.0.23:34255` and `SERVER=10.5.0.7:34256`?

File: docker-compose-config-c.yaml [hunk 3]
```diff
@@ -158,7 +183,26 @@ services:
...
+
+  sv2-tranlator-node-proxy:
+    image: sv1-custom-proxy-builder-image
+    command: ["./target/release/sv1-custom-proxy"]
+    ports:
+      - "34255:34255"
+      - "5676:5676"
+    environment:
+      - SERVER=0.0.0.0:34255
+      - CLIENT=10.5.0.7:34256
```
Comment (GitGab19) [inline docker-compose-config-c.yaml:196]: Same doubt as above

File: grafana/provisioning/dashboards/config-a/Docker Prometheus Monitoring-1571332751387.json
Comment (GitGab19) [inline grafana/provisioning/dashboards/config-a/Docker Prometheus Monitoring-1571332751387.json:1]: Screenshot 2024-07-22 at 10 24 39

Referring to this, screenshot, I would move the SV1 New Job Latency on the left, and I would put SV2 New Job Latency plot on the right.
After them, the same for the New Job with prev_hash (that I would rename in "New Template (after a new block found) Latency"

File: sv1-custom-proxy/src/main.rs [hunk 4]
```diff
@@ -89,6 +91,49 @@ async fn transfer(
...
+                                    .duration_since(std::time::UNIX_EPOCH)
+                                    .expect("Time went backwards")
+                                    .as_millis() as f64;
+                                let prometheus_url = "http://10.5.0.21:4567/metrics";
+                                let client = reqwest::Client::new();
+                                if let Ok(response) = client.get(prometheus_url).send().await {
+                                    if let Ok(body) = response.text().await {
+                                        for line in body.lines() {
+                                            println!("Line: {:?}",line);
+                                            if let Some(start_index) = line.find("prevhash=") {
```
Comment (GitGab19) [inline sv1-custom-proxy/src/main.rs]: Just a reminder to add also the way to measure time to get a new job based on a new prev_hash for the SV1 part

File: sv1-custom-proxy/src/main.rs [hunk 5]
```diff
@@ -238,6 +288,34 @@ async fn handle_rpc_request(
     let body_str = String::from_utf8_lossy(&body_bytes);
     println!("Response: {} {:?}", status, headers);
     println!("Response body: {}", body_str);
+    
+    if let Ok(json) = serde_json::from_slice::<Value>(&body_bytes) {
+        if is_get_block_template {
+            is_get_block_template = false;
+            
+            if let Some(result) = json.get("result") {
```
Comment (GitGab19) [inline sv1-custom-proxy/src/main.rs:324]: Same reminder here

File: sv1-custom-proxy/src/main.rs [hunk 6]
```diff
@@ -353,7 +446,157 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
         if let Err(e) = server.await {
             eprintln!("server error: {}", e);
         }
+    } else if proxy_type == "translator-node" {
```
Comment (GitGab19) [inline sv1-custom-proxy/src/main.rs]: If this is put between translator and miner, I would call it "translator-miner" for better clarity

File: sv1-custom-proxy/src/main.rs [hunk 7]
```diff
@@ -353,7 +446,157 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+        let new_job_time_throught_sv2_jdc =  Arc::new(register_gauge!(
+                    "new_job_jdc_latency",
```
Comment (GitGab19) [inline sv1-custom-proxy/src/main.rs]: I would add a reference to sv2 here, like "new_job_sv2_jdc_latency"

File: sv1-custom-proxy/src/main.rs [hunk 8]
```diff
@@ -353,7 +446,157 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+                    "Time required to complete one tp->jdc , translator->node round of new job"
+                )
+                .unwrap());
+        let new_job_time_throught_sv2_pool =  Arc::new(register_gauge!(
+                    "new_job_pool_latency",
```
Comment (GitGab19) [inline sv1-custom-proxy/src/main.rs]: Same here

File: sv2-custom-proxy/src/main.rs [hunk 9]
```diff
@@ -27,6 +26,10 @@ async fn main() {
     let mut valid_shares: Option<Counter> = None;
     let mut stale_shares: Option<Counter> = None;
     let mut share_submission_timestamp: Option<GaugeVec> = None;
+    let mut share_new_job_timestamp_jdc: Option<GaugeVec> = None;
+    let mut share_new_job_timestamp_pool: Option<GaugeVec> = None;
```
Comment (GitGab19) [inline sv2-custom-proxy/src/main.rs]: Why there's "share" in the name? Isn't it just the new job timestamp?

File: sv2-custom-proxy/src/main.rs [hunk 10]
```diff
@@ -44,6 +47,15 @@ async fn main() {
                 )
                 .unwrap(),
             );
+            share_new_job_timestamp_jdc = Some(
+                register_gauge_vec!(
+                    "share_new_job_timestamp_jdc",
```
Comment (GitGab19) [inline sv2-custom-proxy/src/main.rs]: Same here
And also, I would add a reference to sv2 in the metric name

File: sv2-custom-proxy/src/main.rs [hunk 11]
```diff
@@ -56,6 +68,15 @@ async fn main() {
                 )
                 .unwrap(),
             );
+            share_new_job_timestamp_pool = Some(
+                register_gauge_vec!(
+                    "share_new_job_timestamp_pool",
```
Comment (GitGab19) [inline sv2-custom-proxy/src/main.rs]: Same here

File: grafana/provisioning/dashboards/config-c/Docker Prometheus Monitoring-1571332751387.json
Comment (GitGab19) [inline grafana/provisioning/dashboards/config-c/Docker Prometheus Monitoring-1571332751387.json:1]: Same for config C

File: docker-compose-config-a.yaml (hunk 1 above)
Comment (bit-aloo) [reply docker-compose-config-a.yaml]: I forgot to remove it

File: sv1-custom-proxy/src/main.rs [hunk 12]
```diff
@@ -239,6 +317,40 @@ async fn handle_rpc_request(
...
+                            } else {
+                                flag = "different";
+                            }
+                        }
+                        prev_hash_mutex.lock().unwrap().push_back(prev_hash.clone());
+                        let current_timestamp = std::time::SystemTime
+                            ::now()
+                            .duration_since(std::time::UNIX_EPOCH)
+                            .expect("Time went backwards")
+                            .as_millis() as f64;
```
Comment (GitGab19) [inline sv1-custom-proxy/src/main.rs:341]: I would move this right after the `if let Some(prevhash) = previousblockhash.as_str() {`

File: sv1-custom-proxy/src/main.rs [hunk 13]
```diff
@@ -353,7 +505,176 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+            ).unwrap()
+        );
+        let new_job_time_sv2_jdc = Arc::new(
+            register_gauge!("new_job_jdc_new_template", "new job jdc new template").unwrap()
+        );
+        let new_job_time_sv2_pool = Arc::new(
+            register_gauge!("new_job_pool_new_template", "new job pool new template").unwrap()
+        );
+
+        let listener = tokio::net::TcpListener::bind("0.0.0.0:34255").await.unwrap();
```
Comment (GitGab19) [inline sv1-custom-proxy/src/main.rs:528]: I would read this from ENV variables

File: sv1-custom-proxy/src/main.rs [hunk 14]
```diff
@@ -353,7 +505,176 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+        println!("SV2 proxy translation proxy started at 34255");
+        loop {
+            let (inbound, _) = listener.accept().await.unwrap();
+            let outbound = TcpStream::connect("10.5.0.7:34256").await.unwrap();
```
Comment (GitGab19) [inline sv1-custom-proxy/src/main.rs:532]: Same here

**Pull request merged**